### ISSUE TITLE
vecty.Render into non-body component

### ISSUE THREAD

**Issue by slimsag** [gopherjs/vecty#81]
Currently Render is restricted to a body component; i.e. Vecty must own the whole page effectively. We should allow rendering directly into an element that we own, e.g. `vecty.RenderInto("elem-id")` or something.

**Comment by pdf** [comment]
How smart/flexible do we want this to be, would something like:

```go
func RenderInto(node *js.Object, elem ComponentOrHTML) {}
```
```go
vecty.RenderInto(js.Global.Call("getElementByID", "elem-id"), &MyComponent{})
```

be adequate?  Either way, I don't think this should present any difficulty, unless I'm missing something.

**Comment by slimsag** [comment]
Yep, the above is exactly what I am thinking. It should be trivial to do, I just hadn't gotten around to it.

**Comment by marwan-at-work** [comment]
Hey there :v: -- chiming in here since I've been looking forward to this feature :) 

Do you guys think that `js.Global.Call` is a little too low level for such a nicely designed API? 

In my humble opinion, I think it would be better to do something like 

```golang
func RenderInto(id string, elem ComponentOrHTML)
```

And then have vecty grab the element with that ID. 

This won't let you grab an element by attributes other than ID but I think it's worth it. Another option would be to use Honnef's [dom](https://godoc.org/honnef.co/go/js/dom) package so you can do something like: 

```golang
document := dom.GetWindow().Document()
vecty.RenderInto(document.GetElementByID("blah"), &myComponent{})
```

However, this means a third-party dependency so I understand if it's not considered. 

If you think about it, JavaScript has `window`, and `document` almost as part of the language (except for Node.js obviously). So something like Dominik Honnef's `dom` should be supported on the GopherJS level, but that's a little out of scope .

Thanks to you both for the awesome work :)

**Comment by pdf** [comment]
Hey @marwan-at-work my thinking was that we should provide the freedom for the user to use whatever lib or selection method (`js.Global.Call` is the pure example) they might want to (eg your `dom` option as an alternative), and the only type that can satisfy that is `*js.Object`.  Using `dom` could work something like this (I'm not familiar with `dom`, this is based on a 2-second look at the docs):

```go
document := dom.GetWindow().Document()
vecty.RenderInto(document.GetElementByID("blah").Underlying(), &myComponent{})
```

WDYT?

**Comment by slimsag** [comment]
@marwan-at-work there is some relevant discussion here: https://github.com/gopherjs/vecty/issues/179

I think the short answer is that:

1. I don't want to use the `dom` package because it's simply too large a dependency to pull in for such a small task, and it would mean every Vecty app would have to pull in all `dom` package code to operate, even if they don't use `RenderInto`.
2. Because of pending WebAssembly work, it's not clear whether or not `*js.Object` will work in the long run or if we will need something more generic like an interface, build tags, etc, to support both GopherJS and/or WebAssembly support from Go in the future.

I think the right decision for now is `*js.Object` and (perhaps) if there is enough desire for it:

```
// Selector can be "#id" `.class-name` etc, any valid CSS selector matching one element.
func RenderIntoSelector(selector string, elem ComponentOrHTML)
```

Naming for `RenderIntoSelector` and `RenderInto` TBD if we decide to have both.

**Comment by marwan-at-work** [comment]
I really like the `RenderIntoSelector` idea. if there's a selector parser already in place, then it should be pretty simple to build this function on top of the original `RenderInto` function. 

@pdf the `.Underlying()` workaround is also not bad at all. 

I always thought of `*js.Object` as a low level primitive that is hidden under more Go-like APIs that are then exposed to the end user of your package. This also allows for supporting web assembly later on.

In any case, it's a nice feature to land!

**Comment by pdf** [comment]
I think I'd be fine with:

```go
func RenderInto(selector string, elem ComponentOrHTML) {}
```

Then I think we can just use `js.Global.Call("querySelector", selector)` to get the element? Are there downsides to this, scenarios where users may want to pass in a raw element handle?

Too early to tell what WASM will look like if/when it arrives - there is not even any DOM interop in the spec yet.

**Comment by slimsag** [comment]
One reason that would not be good is if you programatically created an element using the `js` package and wanted Vecty to render into it. In that case, you would have to e.g. give it a random unique ID just to invoke `RenderInto`.

> Too early to tell what WASM will look like if/when it arrives - there is not even any DOM interop in the spec yet.

While it's true there is no DOM interop in the spec yet, there is JS interop, and many languages are already using this for DOM interop (e.x. Rust invokes JS shim code to invoke DOM APIs). I would not be surprised if Go's webassembly support gets something like this shortly after it is released. I agree it's too early to tell, though, simply because we don't know where that package will live (stdlib? golang.org/x/wasm ? etc)

**Comment by slimsag** [comment]
@marwan-at-work 

> if there's a selector parser already in place, then it should be pretty simple to build this function on top of the original RenderInto function.

FYI modern browsers have one built-in:

- https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector
- https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll
- https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelector
- https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelectorAll

**Comment by slimsag** [comment]
If anyone wants to send a PR for this, I think the right approach here is to expose two things:

```Go
// Selector can be "#id" `.class-name` etc, any valid CSS selector matching one element.
func RenderInto(selector string, elem ComponentOrHTML) {
    RenderInto(js.Global.Get("document").Call("querySelector", selector), elem)
}

// Like RenderInto, except it takes an HTML node directly rather than a selector.
func RenderIntoNode(node *js.Object, elem ComponentOrHTML) {
    ... see RenderBody for details on how to implement this ...
}
```

I'd accept a PR that implements those two + has better docs than what I wrote & one test for each of them.

**Comment by slimsag** [comment]
This has just been fixed by #249 and documentation is available here:

https://godoc.org/github.com/gopherjs/vecty#RenderInto

https://godoc.org/github.com/gopherjs/vecty#RenderIntoNode
